runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,226 @@
+         */
+        @Getter
+        final UUID ID;
+
+        @Getter
+        final E entry;
+
+        public String toString() {
+            return String.format(ID.toString()+"=>"+entry.toString());
+        }
+
+        CorfuQueueRecord(UUID id, E entry) {
+            this.ID = id;
+            this.entry = entry;
+        }
+    }
+
+    final private static int MAX_BITS_FOR_INDEX = 24;
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     * This is the primary method of consumption of entries enqueued into CorfuQueue.
+     *
+     * To re-constitute the commit order in the presence of transactions the CorfuRecords returned
+     * here have their UUID fields composed of two parts:
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot|     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries < 0) {
+            maxEntries = Integer.MAX_VALUE;
+        }
+
+        // Bind the iteration order to a snapshot of the Queue using a transaction.
+        long snapshotVersion;
+        boolean startedNewTransaction = false;
+        if (TransactionalContext.isInTransaction()) {
+            snapshotVersion = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
+        } else {
+            runtime.getObjectsView().TXBuild().type(TransactionType.WRITE_AFTER_WRITE)
+                    .build()
+                    .begin();
+            snapshotVersion = TransactionalContext.getCurrentContext().getSnapshotTimestamp().getSequence();
+            startedNewTransaction = true;
+        }
+        List<CorfuQueueRecord<E>> copy = new ArrayList<>(Math.min(corfuTable.size(), maxEntries));
+        int index = 0;
+        for (Map.Entry<Long, E> entry : corfuTable.entrySet()) {
+            if (++index >= maxEntries) {
+                break;
+            }
+            UUID orderedId = new UUID((snapshotVersion<<MAX_BITS_FOR_INDEX) | index,
+            entry.getKey());
+            CorfuQueueRecord<E> record = new CorfuQueueRecord<>(orderedId, entry.getValue());
+            copy.add(record);
+        }
+        // Given that we are using a WRITE_AFTER_WRITE on a read-only txn, we expect no aborts.
+        if (startedNewTransaction) {
```

Comment:
You do not geed state for this functionality. Why not:

```
private entryListImpl() {
  ...
}

public entryList() {

  if (!inTransaction) {
    return entryList();
  }

  startTx();
  entryList();
  endTx();
}
```

Reply:
The transaction is not for the snapshot of the entryList() that is already protected by the VLO.
The transaction is a means of obtaining and binding the snapshot version of the list with its index. I have updated the comments on the Java doc in the latest upload explaining the same.